### DISCUSSION TITLE
Merging a dev branch shows lots of very old commits

### DISCUSSION THREAD

**Discussion by yitam** [community/community#22787] [New to GitHub]
<p>Hi, I’m trying to merge the “dev” branch to “master” branch but when creating a pull request, it contains many, many old commits.</p>
<p>I used to follow the steps outlined in <a href="http://git.tutorialhorizon.com/2014/10/05/replace-the-master-branch-with-another-branch-in-git/" rel="noopener nofollow ugc">http://git.tutorialhorizon.com/2014/10/05/replace-the-master-branch-with-another-branch-in-git/</a> but I wonder if there’s a better way to fix this mess once and for all. Ideally, I just need to create a pull request to base master and compare dev branch on GitHub. Plain and simple. It should technically show the commits since the last time I did the merge, shouldn’t it?</p>
<p>Thanks in advance!</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="yitam" data-post="1" data-topic="136053">
<div class="title">
<div class="quote-controls"></div>
 yitam:</div>
<blockquote>
<p>It should technically show the commits since the last time I did the merge, shouldn’t it?</p>
</blockquote>
</aside>
<p>As long as there haven’t been any rebases or similar actions, yes. For example, if your last merge was a rebase merge it has recreated the merged commits, so they technically aren’t the same, and will be merged again unless you modify your dev branch to be based on the rebased commits. Similar for squashed commits, or any rebases outside Github.</p>

**Comment by yitam** [comment]
<p>Thanks for your reply. I think my question really is “how to make my master branch the same as my dev branch” from now on. Now that the two branches are so different, how should I proceed?</p>
<p>To simply rename these two branches is the last resort, I suppose.</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>Well, it depends on whether you are okay with force-pushing and the associated issues or not. If not, you’ll have to merge and resolve any conflicts in favor of the <code>dev</code> branch.</p>
<p>If you are okay with force-pushing and losing everything that’s different from <code>dev</code> on the <code>master</code> branch, you can use <a href="https://git-scm.com/docs/git-reset#Documentation/git-reset.txt-emgitresetemltmodegtltcommitgt"><code>git reset [&lt;mode&gt;] [&lt;commit&gt;]</code></a>:</p>
<blockquote>
<p>This form resets the current branch head to <code>&lt;commit&gt;</code> and possibly updates the index (resetting it to the tree of <code>&lt;commit&gt;</code>) and the working tree depending on <code>&lt;mode&gt;</code>.</p>
</blockquote>
<p>So, you can do something like this:</p>
<pre><code class="lang-auto">git checkout master
git reset dev
git push --force origin master
</code></pre>
<p>Obligatory warning: If you do this, you will lose anything that’s only on the current <code>master</code>, and everyone who has cloned the repository will get a warning when fetching. :warning:</p>
